### PR TITLE
CLN: some code cleanups in pandas/_libs/

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by MomIsBestFriend** [comment]
The only "big" changes I have done here is taking this pattern:

```python

arr = ["a", "b", "c"]

for i in range(len(arr)):
    print(arr[i])
```

and making it to look like this:

```python
arr = ["a", "b", "c"]

for value in arr:
    print(value)
```
---

When it comes to python, I can't see why would one use the first pattern, but when it comes to Cython, I don't know if there's some sort of optimization, that the first pattern gives that the second one don't

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -246,7 +263,7 @@ def item_from_zerodim(val: object) -> object:
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
-def fast_unique_multiple(list arrays, sort: bool=True):
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx]: for functions that are neither cdef nor cpdef, we can use python-style annotations.

the vague/vain hope is that we can get this code close enough to being valid python that we can use python linters on it

File: pandas/_libs/sparse.pyx [hunk 2]
```diff
@@ -448,7 +448,7 @@ cdef class BlockIndex(SparseIndex):
         ylen = y.blengths
 
         # block may be split, but can't exceed original len / 2 + 1
-        max_len = int(min(self.length, y.length) / 2) + 1
```
Comment (jbrockmendel) [inline pandas/_libs/sparse.pyx]: are there no cases where `int` rounds up?

File: pandas/_libs/tslibs/timezones.pyx [hunk 3]
```diff
@@ -217,7 +217,7 @@ cdef object get_dst_info(object tz):
     if cache_key is None:
         # e.g. pytz.FixedOffset, matplotlib.dates._UTC,
         # psycopg2.tz.FixedOffsetTimezone
-        num = int(get_utcoffset(tz, None).total_seconds()) * 1000000000
+        num = int(get_utcoffset(tz, None).total_seconds()) * 1_000_000_000
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timezones.pyx]: code check for cython hard-coded inters over 7ish characters?

**Comment by jorisvandenbossche** [comment]
> but when it comes to Cython, I don't know if there's some sort of optimization, that the first pattern gives that the second one don't

Then you will need to show that it gives similar results (eg show some timings of the changed code, or look at the generated C code).

For example, with a simple test (but I don't know if that is fully equivalent with the changes you did, this was just a quick test), I see a big overhead when iterating directly through the values of an array. Eg check

```
%%cython -a
import cython

import numpy as np
cimport numpy as np


@cython.wraparound(False)
@cython.boundscheck(False)
def sum1():
    cdef:
        np.int64_t[:] arr
        int n
        int val, total
    
    arr = np.random.randint(10, size=10000)
    
    n = len(arr)
    total = 0
    
    for i in range(n):
        val = arr[i]
        total += val
    
    return total
```

vs

```
%%cython -a
import cython

import numpy as np
cimport numpy as np


@cython.wraparound(False)
@cython.boundscheck(False)
def sum2():
    cdef:
        np.int64_t[:] arr
        np.int64_t val, total
    
    arr = np.random.randint(10, size=10000)
    
    n = len(arr)
    total = 0
    
    for val in arr:
        total += val
    
    return total
```

the second version is much slower (and with the annotation in the notebook you clearly see a lot of python interaction). But I am no cython expert, so I might also have done something wrong.

File: pandas/_libs/tslibs/period.pyx [hunk 4]
```diff
@@ -1463,16 +1469,13 @@ def extract_freq(ndarray[object] values):
     # TODO: Change type to const object[:] when Cython supports that.
 
     cdef:
-        Py_ssize_t i, n = len(values)
-        object p
-
-    for i in range(n):
-        p = values[i]
+        object value
```
Comment (WillAyd) [inline pandas/_libs/tslibs/period.pyx]: As mentioned by @jorisvandenbossche I would be hesitant to make this change unless you can run benchmarks showing no difference; this is going from a platform native type to an object so may miss out on optimization

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/period.pyx [hunk 5]
```diff
@@ -14,15 +12,25 @@ from libc.string cimport strlen, memset
...
-    pandas_datetime_to_datetimestruct, check_dts_bounds,
-    NPY_DATETIMEUNIT, NPY_FR_D, NPY_FR_us)
+    npy_datetimestruct,
+    dtstruct_to_dt64,
+    dt64_to_dtstruct,
+    pandas_datetime_to_datetimestruct,
+    check_dts_bounds,
+    NPY_DATETIMEUNIT,
+    NPY_FR_D,
+    NPY_FR_us
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/period.pyx]: missing trailing commas in some of these

File: pandas/_libs/tslibs/period.pyx [hunk 6]
```diff
@@ -1463,16 +1469,13 @@ def extract_freq(ndarray[object] values):
...
-        p = values[i]
+        object value
 
+    for value in values:
         try:
             # now Timestamp / NaT has freq attr
-            if is_period_object(p):
-                return p.freq
+            if is_period_object(value):
+                return value.freq
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/period.pyx]: +1 for avoiding 1-character variable name

**Comment by jbrockmendel** [comment]
@MomIsBestFriend i'd advocate reverting the non-easy parts of this and getting the clear improvements in

**Comment by MomIsBestFriend** [comment]
@jbrockmendel NP, I will do it over the weekend as soon as I have time.

File: pandas/_libs/sparse.pyx (hunk 2 above)
Comment (MomIsBestFriend) [reply pandas/_libs/sparse.pyx]: No, in order for ```int()``` to round up it needs to be done with ```math.ceil``` or with a "hack" like:

```python
number // divider + (number % divider > 0)
```

---

For example:
```python

number = 42.01
divider = 8

number / divider # 5.25125
number // divider # 5.0
int(number / divider) # 5
number // divider + (number % divider > 0) # 6.0
```

File: pandas/_libs/tslibs/timezones.pyx (hunk 3 above)
Comment (MomIsBestFriend) [reply pandas/_libs/tslibs/timezones.pyx]: We don't have a code check for that, and it's not cython specific, it's just something I see here and there that I like to fix, would you be interested in a code check for that kind of thing?

File: pandas/_libs/tslibs/period.pyx [hunk 7]
```diff
@@ -131,9 +142,7 @@ cdef int64_t get_daytime_conversion_factor(int from_index, int to_index) nogil:
         int col = max_value(from_index, to_index)
     # row or col < 6 means frequency strictly lower than Daily, which
     # do not use daytime_conversion_factors
-    if row < 6:
-        return 0
-    elif col < 6:
+    if row < 6 or col < 6:
```
Comment (jorisvandenbossche) [inline pandas/_libs/tslibs/period.pyx]: Doesn't this need brackets to be done in the correct order?

Comment (MomIsBestFriend) [reply pandas/_libs/tslibs/period.pyx]: I have reverted this, as I was doing it wrong before, I haven't looked at the possibility that row can be let's say 12 and col be 2.

Reverted in 208bc03

File: pandas/_libs/tslibs/period.pyx [hunk 8]
```diff
@@ -289,17 +299,15 @@ cdef int64_t DtoB(npy_datetimestruct *dts, int roll_back,
     return DtoB_weekday(unix_date)
 
 
-cdef inline int64_t upsample_daytime(int64_t ordinal,
-                                     asfreq_info *af_info) nogil:
+cdef inline int64_t upsample_daytime(int64_t ordinal, asfreq_info *af_info) nogil:
     if (af_info.is_end):
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/period.pyx]: if you wanted to remove the parens here, i wouldnt object

File: pandas/_libs/tslibs/strptime.pyx [hunk 9]
```diff
@@ -78,12 +77,10 @@ def array_strptime(object[:] values, object fmt,
     if fmt is not None:
         if '%W' in fmt or '%U' in fmt:
             if '%Y' not in fmt and '%y' not in fmt:
-                raise ValueError("Cannot use '%W' or '%U' without "
-                                 "day and year")
+                raise ValueError("Cannot use '%W' or '%U' without day and year")
             if ('%A' not in fmt and '%a' not in fmt and '%w' not
                     in fmt):
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/strptime.pyx]: the `in fmt` can go on the previous line, then parens removed

File: pandas/_libs/tslibs/timezones.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timezones.pyx]: i think we can find better uses of time.

**Comment by jbrockmendel** [comment]
@MomIsBestFriend can you rebase and ping on green

**Comment by MomIsBestFriend** [comment]
ping @jbrockmendel

**Comment by jbrockmendel** [comment]
thanks @MomIsBestFriend

**Pull request merged**